tests/beta/tapes/test_qubit_param_shift.py

```diff
@@ -0,0 +1,446 @@
+
+
+class TestParameterShiftRule:
+    """Tests for the parameter shift implementation"""
+
+    @pytest.mark.parametrize("theta", np.linspace(-2 * np.pi, 2 * np.pi, 7))
+    @pytest.mark.parametrize("shift", [np.pi / 2, 0.3, np.sqrt(2)])
+    @pytest.mark.parametrize("G", [qml.RX, qml.RY, qml.RZ, qml.PhaseShift])
+    def test_pauli_rotation_gradient(self, mocker, G, theta, shift, tol):
+        """Tests that the automatic gradients of Pauli rotations are correct."""
+        spy = mocker.spy(QubitParamShiftTape, "parameter_shift")
+        dev = qml.device("default.qubit", wires=1)
+
+        with QubitParamShiftTape() as tape:
+            qml.QubitStateVector(np.array([1., -1.]) / np.sqrt(2), wires=0)
+            G(theta, wires=[0])
+            expval(qml.PauliZ(0))
+
+        tape.trainable_params = {1}
+
+        autograd_val = tape.jacobian(dev, shift=shift, method="analytic")
+        manualgrad_val = (
+            tape.execute(dev, params=[theta + np.pi / 2])
+            - tape.execute(dev, params=[theta - np.pi / 2])
+        ) / 2
+        assert np.allclose(autograd_val, manualgrad_val, atol=tol, rtol=0)
+
+        assert spy.call_args[1]["shift"] == shift
+
+        # compare to finite differences
+        numeric_val = tape.jacobian(dev, shift=shift, method="numeric")
+        assert np.allclose(autograd_val, numeric_val, atol=tol, rtol=0)
+
+    @pytest.mark.parametrize("theta", np.linspace(-2 * np.pi, 2 * np.pi, 7))
+    @pytest.mark.parametrize("shift", [np.pi / 2, 0.3, np.sqrt(2)])
+    def test_Rot_gradient(self, mocker, theta, shift, tol):
+        """Tests that the automatic gradient of a arbitrary Euler-angle-parameterized gate is correct."""
+        spy = mocker.spy(QubitParamShiftTape, "parameter_shift")
+        dev = qml.device("default.qubit", wires=1)
+        params = np.array([theta, theta ** 3, np.sqrt(2) * theta])
+
+        with QubitParamShiftTape() as tape:
+            qml.QubitStateVector(np.array([1., -1.]) / np.sqrt(2), wires=0)
+            qml.Rot(*params, wires=[0])
+            expval(qml.PauliZ(0))
+
+        tape.trainable_params = {1, 2, 3}
+
+        autograd_val = tape.jacobian(dev, shift=shift, method="analytic")
+        manualgrad_val = np.zeros_like(autograd_val)
+
+        for idx in list(np.ndindex(*params.shape)):
+            s = np.zeros_like(params)
+            s[idx] += np.pi / 2
+
+            forward = tape.execute(dev, params=params + s)
+            backward = tape.execute(dev, params=params - s)
+
+            manualgrad_val[0, idx] = (forward - backward) / 2
+
+        assert np.allclose(autograd_val, manualgrad_val, atol=tol, rtol=0)
+        assert spy.call_args[1]["shift"] == shift
+
+        # compare to finite differences
+        numeric_val = tape.jacobian(dev, shift=shift, method="numeric")
+        assert np.allclose(autograd_val, numeric_val, atol=tol, rtol=0)
+
+    @pytest.mark.xfail(reason="CRX gate does not satisfy the parameter-shift rule")

```

Comment:
The plan is to:

1. Determine the 3- or 4-term parameter-shift rule that is required for CR gates

2. Modify the `Operation.grad_recipe` attribute and the `parameter_shift` method to allow for operations specifying parameter-shifts via arbitrary linear combinations. That is, instead of `grad_recipe = (multiplier, shift)` (which assumes a 2-term rule), we should have `grad_recipe = [(c0, s0), (c1, s1), ...]` for each term in an arbitrary-term parameter-shift.

This should be relatively easy to do in order to get the CR gates to satisfy the parameter-shift rule.

---

In the meantime, we should do the following:

* *No decomposition:* Define `op.grad_method = "F"` for the `CR` gates, so that _without_ decomposition they will instead request finite differences.

* *Decomposition:* Force `tape.expand()` to decompose any CR gates (already implemented inside the new QNode).